mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp

```diff
@@ -118,22 +133,29 @@ ForLowering::matchAndRewrite(ForOp forOp, OpAdaptor adaptor,
   emitc::ForOp loweredFor = rewriter.create<emitc::ForOp>(
       loc, adaptor.getLowerBound(), adaptor.getUpperBound(), adaptor.getStep());
 
-  // Propagate any attributes from the ODS forOp to the lowered emitc::for op.
-  loweredFor->setAttrs(forOp->getAttrs());
-
   Block *loweredBody = loweredFor.getBody();
 
   // Erase the auto-generated terminator for the lowered for op.
   rewriter.eraseOp(loweredBody->getTerminator());
 
+  // Convert the original region types into the new types by adding unrealized
+  // casts in the beginning of the loop. This performs the conversion in place.
```

Comment:
I'm wondering if this doesn't also convert the return types of the block, thereby resulting in type mismatches for the yield op. Since `lowerYield` also does that, either `convertRegionTypes` only converts the block arguments, or it returns inconsistent IR with yield types that are different from the block return types. Do you have a better insight on this than I do?

Reply:
The `rewriter.convertRegionTypes()` only converts the arguments and the SSA value returned by the yield must not be listed in the argument list of the region. The yield and the value being return is either unchanged or the result of a new unrealized cast (emitc.size_t -> index). So the type mismatch is still there after this function is called. 

Later on, as we lower the `yield` in `lowerYield` we perform the conversion to emitc (by introducing the new `emitc.assign`). I think this happens mostly because of the result variables of the `scf.for` loop that are lowered here. If this new `emitc.assign` does not have the right type (before it had RHS: `emitc.size_t` LHS: `index`, now it has RHS: `emitc.size_t` LHS: `emitc.size_t`) it doesn't seem to be converted while we convert the body of the loop to emitc. That I didn't look deeper but maybe that happens because the assign is now on the emitc dialect and not the original scf.dialect.

Reply:
(Even after the last changes to the PR, what I said still holds. The only difference is that the introduction of any reinterpret_casts is carried by the `ConverterRewriterPattern`)